### PR TITLE
Improve perfromance for loading from Stream on Windows

### PR DESCRIPTION
**Pull request by JeremyKuhne**
- Make IStream blittable and move to Interop
- Change GPStream to use Span APIs
- Make STATSTG blittable and move to Interop
- Fill out corners correctly in GPStream
- Remove extra allocations opening Gifs
- Fix allocations in FrameDimensionsList and move to shared
- Move IPicture to Icon.Windows (was Windows specific and used IStream, needs cleanup and moving to Interop)
- Add perf tests for opening images

### CODE REVIEW

**Comment by JeremyKuhne** [comment]
Before:

```
  System.Drawing.Common.Performance.Tests.dll                            | Metric                                        | Unit  | Iterations |    Average | STDEV.S |        Min |        Max
  :---------------------------------------------------------------------- |:--------------------------------------------- |:-----:|:----------:| ----------:| -------:| ----------:| ----------:
   System.Drawing.Tests.Perf_Image_Load.Bitmap_FromStream                 | Duration                                      | msec  |     5      |   2104.315 |  47.190 |   2052.298 |   2170.079
   System.Drawing.Tests.Perf_Image_Load.Bitmap_FromStream                 | Allocation Size on Benchmark Execution Thread | bytes |     5      | 1.207E+009 |   0.000 | 1.207E+009 | 1.207E+009
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream                  | Duration                                      | msec  |     5      |   2061.762 |  33.282 |   2039.791 |   2118.484
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream                  | Allocation Size on Benchmark Execution Thread | bytes |     5      | 1.207E+009 |   0.000 | 1.207E+009 | 1.207E+009
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation     | Duration                                      | msec  |    100     |     11.951 |   0.623 |     11.371 |     15.602
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation     | Allocation Size on Benchmark Execution Thread | bytes |    100     | 213600.000 |   0.000 | 213600.000 | 213600.000
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation_Gif | Duration                                      | msec  |    100     |     20.703 |   0.606 |     19.907 |     22.184
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation_Gif | Allocation Size on Benchmark Execution Thread | bytes |    100     | 5.438E+006 |   0.000 | 5.438E+006 | 5.438E+006
```

After:

```
   System.Drawing.Common.Performance.Tests.dll                            | Metric                                        | Unit  | Iterations |   Average | STDEV.S |       Min |       Max
  :---------------------------------------------------------------------- |:--------------------------------------------- |:-----:|:----------:| ---------:| -------:| ---------:| ---------:
   System.Drawing.Tests.Perf_Image_Load.Bitmap_FromStream                 | Duration                                      | msec  |     6      |  1883.708 |  13.440 |  1870.467 |  1902.180
   System.Drawing.Tests.Perf_Image_Load.Bitmap_FromStream                 | Allocation Size on Benchmark Execution Thread | bytes |     6      | 31200.000 |   0.000 | 31200.000 | 31200.000
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream                  | Duration                                      | msec  |     6      |  1879.785 |   4.414 |  1874.196 |  1885.028
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream                  | Allocation Size on Benchmark Execution Thread | bytes |     6      | 31200.000 |   0.000 | 31200.000 | 31200.000
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation     | Duration                                      | msec  |    100     |    12.245 |   0.301 |    11.442 |    13.866
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation     | Allocation Size on Benchmark Execution Thread | bytes |    100     | 31200.000 |   0.000 | 31200.000 | 31200.000
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation_Gif | Duration                                      | msec  |    100     |    19.631 |   3.527 |    18.657 |    54.310
   System.Drawing.Tests.Perf_Image_Load.Image_FromStream_NoValidation_Gif | Allocation Size on Benchmark Execution Thread | bytes |    100     | 10400.000 |   0.000 | 10400.000 | 10400.000
```

A key thing to note is that most of the time spent opening images is validating them. On this particular machine (which is loading from NVM/e), validating is 170 times slower. Other measures I've seen are in the 150-200 times slower. We should look at adding the option to skip and flipping the default behavior for 3.0 (with an AppContext switch to force).

Unix can be improved as well. @safern plans to investigate.

**Comment by JeremyKuhne** [comment]
@dotnet-bot test Windows x64 Debug Build

**Comment by safern** [comment]
Thanks for doing this. Love the results.

The reason is because the package needs to be netcoreapp2.0 compatible and the netstandard implementation is a not supported assembly. However I think it would be fine now that we’ve shipped the compat pack to harvest the 2.0 assets.

@weshaggard any complains with doing that?

**Comment by ericstj** [comment]
If you consider the netstandard2.0 surface area done then go ahead and harvest all the old implementations and refs.

If you imagine that you still might want to grow netstandard2.0 surface area (either adding stuff already in desktop or new types) then you should keep those builds around and ifdef.

**Comment by safern** [comment]
> If you consider the netstandard2.0 surface area done then go ahead and harvest all the old implementations and refs.
> If you imagine that you still might want to grow netstandard2.0 surface area (either adding stuff already in desktop or new types) then you should keep those builds around and ifdef.

We're going to be doing a lot of this perf work which I think it will involve using Span and those ref-like 2.1 apis, so I think it is not worth the effort of having if defs all over the place for this perf effort only to support ns2.0, if we add new types/apis I think it would be fine for people to move to vnext.

**Comment by ericstj** [comment]
>  if we add new types/apis I think it would be fine for people to move to vnext.

You can not make that generalization for all packages.  It's not "move to vNext" for many customers it's "target the majority of live frameworks" to "target a single netcoreapp3.0 framework".  That's not a small thing and for many customers will be a deal breaker.  You'll need to decide that based on the business case for APIs being added.  For things built on top of netstandard you need to remember that you are not only serving netcoreapp customers, but also those which want to cross target.

I will repeat: 
1. If you know of API you want to add now that needs to be available for everyone then keep this library cross-compiling.
2. If you don't know of any API that will be added then you can opportunistically harvest to save yourself some work now, you just may need to bring it back in the future.

**Comment by weshaggard** [comment]
I have not issues with harvesting the older assets, but please be explicit about it as @ericstj mentioned. If we add any additional APIs to the netstandard20 ref then you will have to start building the netcoreapp2.0 asset live again in order for it to be compliant with the ref.

**Comment by safern** [comment]
Also, I was looking to System.Drawing.Common configurations.props and I've noticed that we aren't including netcoreapp vnext in the packageconfigurations. If we're going to if def implementation I think we should include that configuration as a packageconfiguration as well so that it takes effect on the package also when using it in vnext. @weshaggard, @ericstj correct me if I'm wrong.

File: src/System.Drawing.Common/src/System/Drawing/Image.cs [hunk 1]
```diff
@@ -314,24 +301,57 @@ internal void SetNativeImage(IntPtr handle)
...
-                foreach (Guid guid in dimensions)
+                Gdip.CheckStatus(Gdip.GdipImageGetFrameDimensionsCount(new HandleRef(image, image.nativeImage), out int dimensions));
+                if (dimensions <= 0)
                 {
-                    FrameDimension dimension = new FrameDimension(guid);
-                    if (dimension.Equals(FrameDimension.Time))
+                    return;
+                }
+
+                Guid* guids = stackalloc Guid[dimensions];
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Image.cs]: do we know that the dimensions are small enough to always stackalloc?

Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Image.cs]: I'm not sure, I'll update to fall back to the array if crazy.

**Comment by ericstj** [comment]
> Also, I was looking to System.Drawing.Common configurations.props and I've noticed that we aren't including netcoreapp vnext in the packageconfigurations. If we're going to if def implementation I think we should include that configuration as a packageconfiguration as well so that it takes effect on the package also when using it in vnext. @weshaggard, @ericstj correct me if I'm wrong.

Technically it could be a `netcoreapp2.1` configuration.  Yeah it is needed.  Good catch

File: src/Common/src/Interop/Windows/Interop.HRESULT.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+internal static partial class Interop
+{
+    // https://msdn.microsoft.com/en-us/library/cc231198.aspx
+    internal enum HRESULT : int
+    {
+        S_OK = 0,
```
Comment (danmosemsft) [inline src/Common/src/Interop/Windows/Interop.HRESULT.cs]: Nit, some of these are also defined in Gdiplus.cs

File: src/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 3]
```diff
@@ -11,7 +11,7 @@
...
     <IsPartialFacadeAssembly Condition="'$(TargetsNetFx)' == 'true'">true</IsPartialFacadeAssembly>
     <GeneratePlatformNotSupportedAssemblyMessage Condition="'$(TargetGroup)' == 'netstandard'">SR.PlatformNotSupported_Drawing</GeneratePlatformNotSupportedAssemblyMessage>
-    <Configurations>net461-Debug;net461-Release;netcoreapp-Unix-Debug;netcoreapp-Unix-Release;netcoreapp-Windows_NT-Debug;netcoreapp-Windows_NT-Release;netcoreapp2.0-Unix-Debug;netcoreapp2.0-Unix-Release;netcoreapp2.0-Windows_NT-Debug;netcoreapp2.0-Windows_NT-Release;netfx-Debug;netfx-Release;netstandard-Debug;netstandard-Release</Configurations>
+    <Configurations>net461-Debug;net461-Release;netcoreapp-Unix-Debug;netcoreapp-Unix-Release;netcoreapp-Windows_NT-Debug;netcoreapp-Windows_NT-Release;netcoreapp2.0-Unix-Debug;netcoreapp2.0-Unix-Release;netcoreapp2.0-Windows_NT-Debug;netcoreapp2.0-Windows_NT-Release;netcoreapp2.1-Unix-Debug;netcoreapp2.1-Unix-Release;netcoreapp2.1-Windows_NT-Debug;netcoreapp2.1-Windows_NT-Release;netfx-Debug;netfx-Release;netstandard-Debug;netstandard-Release</Configurations>
   </PropertyGroup>
   <ItemGroup Condition="$(TargetGroup.StartsWith('netcoreapp'))">
```
Comment (ericstj) [inline src/System.Drawing.Common/src/System.Drawing.Common.csproj]: Existing nit, could be `$(TargetsNetCoreApp)`.  Moreover we could probably do a pass of the whole repo to clean up targeting conditions in projects.

Comment (safern) [reply src/System.Drawing.Common/src/System.Drawing.Common.csproj]: I guess it could be in a separate PR, but seems like a nice cleanup.

Comment (ericstj) [reply src/System.Drawing.Common/src/System.Drawing.Common.csproj]: Yeah, I think we could even avoid string comparisons to 'true' as well.

File: src/Common/src/Interop/Windows/Interop.HRESULT.cs (hunk 2 above)
Comment (JeremyKuhne) [reply src/Common/src/Interop/Windows/Interop.HRESULT.cs]: Yeah, plan to get them all cleaned up.

File: src/System.Drawing.Common/src/System/Drawing/Image.cs [hunk 4]
```diff
@@ -314,24 +301,70 @@ internal void SetNativeImage(IntPtr handle)
...
+
+                Span<Guid> guids;
+                if (dimensions < 16)
+                {
+                    Guid* g = stackalloc Guid[dimensions];
+                    guids = new Span<Guid>(g, dimensions);
+                }
+                else
+                {
+                    guids = new Span<Guid>(new Guid[dimensions]);
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Image.cs]: Does 2.0 has the implicit T[] to Span<T> cast? Can't we do: `guids = new Guid[dimensions]`? or is that only at declaration? Same question for the previous assignment.

**Comment by safern** [comment]
@ericstj I wonder why the Packaging All Configurations is failing when building for netcoreapp2.1 if the references are conditioned to `TargetsNetFx != true`

**Comment by ericstj** [comment]
I bet we’re missing the external project to restore 2.1 references

**Comment by safern** [comment]
> I bet we’re missing the external project to restore 2.1 references

So I guess the infrastructure is not ready to build for netcoreapp2.1 -- should we then leave the configuration as netcoreapp and add a comment that whenever we map netcoreapp to 3.0 it should be updated to be netcoreapp2.1 in this project?

**Comment by saucecontrol** [comment]
> A key thing to note is that most of the time spent opening images is validating them. On this particular machine (which is loading from NVM/e), validating is 170 times slower. Other measures I've seen are in the 150-200 times slower range.

To be clear, the `validateImageData` parameter actually controls whether the image is eagerly decoded on load.  If loading only for the purpose of reading metadata or if only a portion of the pixel data is going to be consumed, the perf gain holds up.  If the image would have been fully decoded later anyway, the perf is a wash... just delayed.

> We should look at adding the option to skip (it is only available on one overload right now) and flipping the default behavior for 3.0 (with an AppContext switch to force).

If there's a change to the API, it would be nice if that parameter were renamed to more accurately reflect its actual behavior.  The preferred behavior is difficult to discover with the current API and parameter name.

**Comment by JeremyKuhne** [comment]
> To be clear, the validateImageData parameter actually controls whether the image is eagerly decoded on load.

Thanks! I hadn't gotten a chance to dig all the way through it yet.

> If there's a change to the API, it would be nice if that parameter were renamed to more accurately reflect its actual behavior.

Might be stuck as it was exposed already. :/ At the very least I'll make sure we get the docs updated to be explicit about this. 

If you have anything else you know about System.Drawing/GDI+ related to perf (or anything else) please let us know! I'm only marginally familiar with it- my interest was sparked by noticing how inefficient the interop calls are.

File: src/System.Drawing.Common/src/System/Drawing/Image.cs (hunk 4 above)
Comment (JeremyKuhne) [reply src/System.Drawing.Common/src/System/Drawing/Image.cs]: Can't in 2.1 as the compiler blocks it. This is the evil workaround. :P

**Comment by saucecontrol** [comment]
> If you have anything else you know about System.Drawing/GDI+ related to perf (or anything else) please let us know!

I did a series of blog posts covering System.Drawing/GDI+ a few years back, including a whole one on that odd ["validation"](https://photosauce.net/blog/post/image-scaling-with-gdi-part-5-push-vs-pull-and-image-validation) behavior.  Most of the stuff I wrote about relates to GDI+ internals that can't be fixed on the managed side of things, but it would be really nice to see the docs made more accurate or some of the defaults changed if that's possible.

File: src/System.Drawing.Common/src/System/Drawing/Internal/GPStream.cs [hunk 5]
```diff
@@ -2,279 +2,239 @@
...
-                    }
-
-                    readLen = stream.Read(bytes, current, ReadBlock);
-                    current += readLen;
-                } while (readLen != 0);
-
-                dataStream = new MemoryStream(bytes);
+                // Copy to a memory stream so we can seek
+                MemoryStream memoryStream = new MemoryStream();
+                _dataStream.CopyTo(memoryStream);
```
Comment (safern) [inline src/System.Drawing.Common/src/System/Drawing/Internal/GPStream.cs]: This should be `stream.CopyTo(memoryStream);`

**Comment by JeremyKuhne** [comment]
@dotnet-bot test Windows x64 Debug Build

@chcosta How can we find these logs? Been seeing this a few times recently.

```
20:25:45 MSBUILD : error MSB4166: Child node "4" exited prematurely. Shutting down. Diagnostic information may be found in files in "C:\Users\dotnet-bot\AppData\Local\Temp\" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory.
```

**Comment by danmosemsft** [comment]
The real fix is https://github.com/Microsoft/msbuild/issues/1038

@mattgal explained to me in the past that there is away to configure our CI to archive arbitrary files, not sure whether it can accept a wildcard pattern though.

The issue is usually out of memory/disk or some task throwing on the threadpool thread.

**Comment by MattGal** [comment]
@danmosemsft I know how to do that on VSTS and we've done this (Via VSTS logging commands) ; I know Jenkins can do this but I wouldn't know offhand.  

If we're talking vsts, this page is all you need to know: https://github.com/Microsoft/vsts-tasks/blob/master/docs/authoring/commands.md

**Pull request merged**